python/lsst/pipe/base/graph/_loadHelpers.py

```diff
@@ -0,0 +1,310 @@
+        return the requested byte range. This is the most flexible
+        implementation, as no special read is required. This will not give a
+        speed up with any sub graph reads though.
+        """
+        if not hasattr(self, 'buffer'):
+            self.buffer = self.uriObject.read()
+        return self.buffer[start:stop]
+
+    def close(self):
+        """Cleans up an instance if needed. Base class does nothing
+        """
+        pass
+
+
+@register_helper(ButlerS3URI)
+class S3LoadHelper(DefaultLoadHelper):
+    # This subclass implements partial loading of a graph using a s3 uri
+    def _readBytes(self, start: int, stop: int) -> bytes:
+        args = {}
+        # minus 1 in the stop range, because this header is inclusive rather
+        # than standard python where the end point is generally exclusive
+        args["Range"] = f"bytes={start}-{stop-1}"
+        try:
+            response = self.uriObject.client.get_object(Bucket=self.uriObject.netloc,
+                                                        Key=self.uriObject.relativeToPathRoot,
+                                                        **args)
+        except (self.uriObject.client.exceptions.NoSuchKey,
+                self.uriObject.client.exceptions.NoSuchBucket) as err:
+            raise FileNotFoundError(f"No such resource: {self}") from err
+        body = response["Body"].read()
+        response["Body"].close()
+        return body
+
+
+@register_helper(ButlerFileURI)
+class FileLoadHelper(DefaultLoadHelper):
+    # This subclass implements partial loading of a graph using a file uri
+    def _readBytes(self, start: int, stop: int) -> bytes:
+        if not hasattr(self, 'fileHandle'):
+            self.fileHandle = open(self.uriObject.ospath, 'rb')
+        self.fileHandle.seek(start)
+        return self.fileHandle.read(stop-start)
+
+    def close(self):
+        if hasattr(self, 'fileHandle'):
+            self.fileHandle.close()
+
+
+@register_helper(io.IOBase)  # type: ignore
+class OpenFileHandleHelper(DefaultLoadHelper):
+    # This handler is special in that it does not get initialized with a
+    # ButlerURI, but an open file handle.
+
+    # Most everything stays the same, the variable is even stored as uriObject,
+    # because the interface needed can stay the same. Unfortunately because we
+    # do not have Protocols yet, this can not be nicely expressed with typing.
+
+    # This helper does support partial loading
+
+    def __init__(self, uriObject: io.IO[bytes]):
+        # This is actually a io handle, but is named uri because the follow the
+        # same api, and this can be used in place of a URI object in this
+        # class.
+        self.uriObject = uriObject
+
+        sizeBytes = self.uriObject.read(18)
+        save_version, taskDefSize, nodeSize = struct.unpack('>HQQ', sizeBytes)
+        self.save_version = save_version
+        self.headerSize = 18+taskDefSize+nodeSize
```

Comment:
This code duplication and that you don't call base class constructor suggests that some refactoring may be in order.